### PR TITLE
[12.0] [MIG] Base User Role History - Migration from 9.0 to 12.0

### PR DESCRIPTION
**Pull request by baimont**
This PR is the migration of this https://github.com/OCA/server-tools/pull/1683

### CODE REVIEW

File: base_user_role_history/tests/test_base_user_role_history.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
+# -*- coding: utf-8 -*-
```
Comment (sebalix) [inline base_user_role_history/tests/test_base_user_role_history.py]: No need to declare encoding starting from Odoo 11 (Python >= 3.5)

File: base_user_role_history/models/res_users.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        for rec in self:
+            role_line_values_by_user.setdefault(rec, {})
+            for role_line in rec.role_line_ids:
+                role_line_values_by_user[rec][role_line.id] = \
+                    self._prepare_role_line_history_dict(role_line)
+        return role_line_values_by_user
+
+    @api.model
+    def create(self, vals):
+        res = super(ResUsers, self).create(vals)
```
Comment (sebalix) [inline base_user_role_history/models/res_users.py]:
```suggestion
        res = super().create(vals)
```

File: base_user_role_history/models/res_users.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    def _get_role_line_values_by_user(self):
+        role_line_values_by_user = {}
+        for rec in self:
+            role_line_values_by_user.setdefault(rec, {})
+            for role_line in rec.role_line_ids:
+                role_line_values_by_user[rec][role_line.id] = \
+                    self._prepare_role_line_history_dict(role_line)
+        return role_line_values_by_user
+
+    @api.model
```
Comment (sebalix) [inline base_user_role_history/models/res_users.py]: Better using `@api.model_create_multi` starting from Odoo >= 12.0 to receive a list of values, and rename `vals` to `vals_list` (https://github.com/odoo/odoo/blob/12.0/odoo/models.py#L3471)

File: base_user_role_history/models/res_users.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            {},
+            new_role_line_values_by_user
+        )
+        res.last_role_line_modification = fields.Datetime.now()
+        return res
+
+    @api.multi
+    def write(self, vals):
+        if 'role_line_ids' not in vals:
+            return super(ResUsers, self).write(vals)
```
Comment (sebalix) [inline base_user_role_history/models/res_users.py]:
```suggestion
            return super().write(vals)
```

File: base_user_role_history/models/res_users.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        old_role_line_values_by_user = self._get_role_line_values_by_user()
+        res = super(ResUsers, self).write(vals)
```
Comment (sebalix) [inline base_user_role_history/models/res_users.py]:
```suggestion
        res = super().write(vals)
```

**Review by sebalix — COMMENTED** [review]
Same small remarks than in https://github.com/OCA/server-tools/pull/1683 :) + Some changes related to Odoo 12

**Comment by baimont** [comment]
> Same small remarks than in [OCA/server-tools#1683](https://github.com/OCA/server-tools/pull/1683) :) + Some changes related to Odoo 12

Hello again @sebalix and thanks again for your useful comments.
They all seemed legit to me, so now this pr takes all your comments into account.

**Review by ThomasBinsfeld — APPROVED** [review]

**Review by sebalix — APPROVED** [review]

**Comment by baimont** [comment]
Hey @sebalix 
could this one be merged?

**Pull request merged**